File: pfe-java/oauth/app/controllers/oauth/OAuth.java
package controllers.oauth;

import com.fasterxml.jackson.databind.JsonNode;
import play.libs.F;
import play.libs.Scala;
import play.libs.ws.WSClient;
import play.mvc.Call;
import play.mvc.Controller;
import play.mvc.Http;
import play.mvc.Result;
import url.URL;

import javax.inject.Inject;
import javax.inject.Singleton;

import java.util.function.Function;
import java.util.function.Supplier;

import static url.URL.param;

@Singleton
public class OAuth extends Controller {

    public static class Configuration {
        public final String authorizationEndpoint;
        public final String tokenEndpoint;
        public final String clientId;
        public final String clientSecret;
        public final String scope;

        public final String tokenKey;
        public final String defaultReturnUrl;

        public Configuration(String authorizationEndpoint, String tokenEndpoint, String clientId, String clientSecret, String scope, String defaultReturnUrl, String tokenKey) {
            this.authorizationEndpoint = authorizationEndpoint;
            this.tokenEndpoint = tokenEndpoint;
            this.clientId = clientId;
            this.clientSecret = clientSecret;
            this.scope = scope;
            this.tokenKey = tokenKey;
            this.defaultReturnUrl = defaultReturnUrl;
        }

        public Configuration(String authorizationEndpoint, String tokenEndpoint, String clientId, String clientSecret, String scope, String defaultReturnUrl) {
            this(authorizationEndpoint, tokenEndpoint, clientId, clientSecret, scope, defaultReturnUrl, "oauth-token");
        }
    }

    final Configuration configuration;
    final WSClient ws;

    @Inject
    public OAuth(WSClient ws, Configuration configuration) {
        this.ws = ws;
        this.configuration = configuration;
    }

    public String authorizeUrl(Call returnTo) {
        return URL.build(configuration.authorizationEndpoint, Scala.varargs(
                param("response_type", "code"),
                param("client_id", configuration.clientId),
                param("redirect_uri", routes.OAuth.callback().absoluteURL(request())),
                param("scope", configuration.scope),
                param("state", returnTo.url())
        ));
    }

    public F.Promise<Result> callback() {
        String code = request().getQueryString("code");
        if (code != null) {
            String state = request().getQueryString("state");
            String returnTo = state != null ? state : configuration.defaultReturnUrl;
            return ws.url(configuration.tokenEndpoint)
                    .setContentType(Http.MimeTypes.FORM)
                    .post(URL.encode(Scala.varargs(
                            param("code", code),
                            param("client_id", configuration.clientId),
                            param("client_secret", configuration.clientSecret),
                            param("redirect_uri", routes.OAuth.callback().absoluteURL(request())),
                            param("grant_type", "authorization_code")
                    ))).map(response -> {
                        JsonNode accessTokenJson = response.asJson().get("access_token");
                        if (accessTokenJson == null || !accessTokenJson.isTextual()) {
                            return internalServerError();
                        } else {
                            String accessToken = accessTokenJson.asText();
                            session().put(configuration.tokenKey, accessToken);
                            return redirect(returnTo);
                        }
                    });
        } else {
            return F.Promise.pure(internalServerError());
        }
    }

    // Must be called within the scope of an Action
    public <A> A authenticated(Function<String, A> f, Supplier<A> g) {
        String token = session(configuration.tokenKey);
        if (token != null) return f.apply(token);
        else return g.get();
    }

}

File: pfe-scala/oauth/app/controllers/oauth/OAuth.scala
package controllers.oauth

import javax.inject.{Inject, Singleton}

import play.api.libs.ws.WSClient
import play.api.mvc.{Action, Call, Controller, RequestHeader}

import scala.concurrent.{ExecutionContext, Future}

@Singleton class OAuth @Inject() (ws: WSClient, val configuration: OAuth.Configuration)(implicit ec: ExecutionContext) extends Controller {

  val callback = Action.async { implicit request =>
    request.getQueryString("code") match {
      case Some(code) =>
//        val returnTo = request.getQueryString("state") getOrElse routes.Items.list().url
        val returnTo = request.getQueryString("state") getOrElse configuration.defaultReturnUrl
        for {
          response <- ws.url(configuration.tokenEndpoint).post(Map(
            "code" -> Seq(code),
            "client_id" -> Seq(configuration.clientId),
            "client_secret" -> Seq(configuration.clientSecret),
            "redirect_uri" -> Seq(routes.OAuth.callback().absoluteURL()),
            "grant_type" -> Seq("authorization_code")
          ))
        } yield {
          (response.json \ "access_token").validate[String].fold(
            _ => InternalServerError,
            token => Redirect(returnTo).addingToSession(configuration.tokenKey -> token)
          )
        }
      case None =>
        Future.successful(InternalServerError)
    }
  }

  def authorizeUrl(returnTo: Call)(implicit request: RequestHeader): String =
    OAuth.makeUrl(configuration.authorizationEndpoint,
      "response_type" -> "code",
      "client_id" -> configuration.clientId,
      "redirect_uri" -> routes.OAuth.callback().absoluteURL(),
      "scope" -> configuration.scope,
      "state" -> returnTo.url
    )

  def authenticated[A](f: String => A, g: => A)(implicit request: RequestHeader): A =
    request.session.get(configuration.tokenKey) match {
      case Some(token) => f(token)
      case None => g
    }

}

object OAuth {

  case class Configuration @Inject() (
    authorizationEndpoint: String,
    tokenEndpoint: String,
    clientId: String,
    clientSecret: String,
    scope: String,
    tokenKey: String = "oauth-token",
    defaultReturnUrl: String = "/")

  def makeUrl(endpoint: String, qs: (String, String)*): String = {
    import java.net.URLEncoder.{encode => enc}
    val params = for ((n, v) <- qs) yield s"""${enc(n, "utf-8")}=${enc(v, "utf-8")}"""
    endpoint + params.toSeq.mkString("?", "&", "")
  }

}